tools/deploy/helper.go

```diff
@@ -0,0 +1,442 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package deploy
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/services/placement"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+	"github.com/m3db/m3x/sync"
+)
+
+var (
+	errNoDeploymentProgress = errors.New("no deployment progress")
+	errInvalidRevision      = errors.New("invalid revision")
+)
+
+// Mode is the deployment mode.
+type Mode int
+
+// A list of supported deployment modes.
+const (
+	DryRunMode Mode = iota
+	ForceMode
+)
+
+// Helper is a helper class handling deployments.
+type Helper interface {
+	// Deploy deploys a target revision to the instances in the placement.
+	Deploy(revision string, mode Mode) error
+}
+
+// TODO(xichen): disable deployment while another is ongoing.
+type helper struct {
+	logger             xlog.Logger
+	planner            planner
+	client             aggregatorClient
+	mgr                Manager
+	store              kv.Store
+	retrier            xretry.Retrier
+	foreverRetrier     xretry.Retrier
+	workers            xsync.WorkerPool
+	toPlacementIDFn    ToPlacementInstanceIDFn
+	toAPIEndpointFn    ToAPIEndpointFn
+	placementWatcher   services.StagedPlacementWatcher
+	settleBetweenSteps time.Duration
+}
+
+// NewHelper creates a new deployment helper.
+func NewHelper(opts HelperOptions) (Helper, error) {
+	client := newAggregatorClient(opts.HTTPClient())
+	planner := newPlanner(client, opts.PlannerOptions())
+	placementWatcherOpts := opts.StagedPlacementWatcherOptions()
+	placementWatcher := placement.NewStagedPlacementWatcher(placementWatcherOpts)
+	if err := placementWatcher.Watch(); err != nil {
+		return nil, err
+	}
+	retryOpts := opts.RetryOptions()
+	retrier := xretry.NewRetrier(retryOpts)
+	foreverRetrier := xretry.NewRetrier(retryOpts.SetForever(true))
+	return helper{
+		logger:             opts.InstrumentOptions().Logger(),
+		planner:            planner,
+		client:             client,
+		mgr:                opts.Manager(),
+		store:              opts.KVStore(),
+		retrier:            retrier,
+		foreverRetrier:     foreverRetrier,
+		workers:            opts.WorkerPool(),
+		toPlacementIDFn:    opts.ToPlacementInstanceIDFn(),
+		toAPIEndpointFn:    opts.ToAPIEndpointFn(),
+		placementWatcher:   placementWatcher,
+		settleBetweenSteps: opts.SettleDurationBetweenSteps(),
+	}, nil
+}
+
+func (h helper) Deploy(revision string, mode Mode) error {
+	if revision == "" {
+		return errInvalidRevision
+	}
+	all, err := h.allInstanceMetadatas()
+	if err != nil {
+		return fmt.Errorf("unable to get all instance metadatas: %v", err)
+	}
+	filtered := filterByRevision(all, revision)
+	plan, err := h.planner.GeneratePlan(filtered, all)
```

Comment:
A general suggestion: It'd be more robust to treat leader->follower elections as common place occurrences, even if they're not going to happen frequently. For e.g. instead of generating the entire plan up front, it'd be better if we generated a step of the plan, executed it, then generated the next step, etc..